File: dotnet/Value/trunk/tmp-origin/main/java/org/ppwcode/value_III/time/interval/AbstractTimeInterval.java
/*<license>
Copyright 2004 - $Date$ by PeopleWare n.v..

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
</license>*/

package org.ppwcode.value_III.time.interval;


import static org.ppwcode.metainfo_I.License.Type.APACHE_V2;
import static org.ppwcode.value_III.time.interval.TimeIntervalRelation.EQUALS;
import static org.ppwcode.value_III.time.interval.TimeIntervalRelation.timeIntervalRelation;

import java.util.Date;

import org.ppwcode.metainfo_I.Copyright;
import org.ppwcode.metainfo_I.License;
import org.ppwcode.metainfo_I.vcs.SvnInfo;
import org.ppwcode.value_III.time.Duration;
import org.ppwcode.vernacular.value_III.AbstractImmutableValue;


/**
 * General supporting code for time interval implementations.
 *
 * @author Jan Dockx
 * @author Peopleware n.v.
 */
@Copyright("2008 - $Date$, PeopleWare n.v.")
@License(APACHE_V2)
@SvnInfo(revision = "$Revision$",
         date     = "$Date$")
public abstract class AbstractTimeInterval extends AbstractImmutableValue implements TimeInterval {

  @Override
  public final boolean equals(Object other) {
//    if (other != null && other instanceof TimeInterval && getBegin() != null &&
//        getEnd() != null && ((TimeInterval)other).getBegin() != null &&
//        ((TimeInterval)other).getEnd() != null) {
//      GregorianCalendar gcB1 = new GregorianCalendar();
//      gcB1.setTime(getBegin());
//      GregorianCalendar gcE1 = new GregorianCalendar();
//      gcE1.setTime(getEnd());
//      GregorianCalendar gcB2 = new GregorianCalendar();
//      gcB2.setTime(((TimeInterval)other).getBegin());
//      GregorianCalendar gcE2 = new GregorianCalendar();
//      gcE2.setTime(((TimeInterval)other).getEnd());
//      if ((gcB1.get(Calendar.YEAR) == 2000)  && (gcE1.get(Calendar.YEAR) == 2122) &&
//          (gcB2.get(Calendar.YEAR) == 2000) && (gcE2.get(Calendar.YEAR) == 2122) &&
//          ((TimeInterval)other).getDuration() != null) {
//        System.out.println("IT'S A HIT");
//      }
//    }
    /* we cannot call super.equals(other) in this case; that checks whether
     * other has the same class as this. In this case that will never happen,
     * since we are abstract. Furthermore we want to compare all TimeIntervals
     * to each other, via the interface.
     */
    return (other != null) && (getClass().isInstance(other)) && timeIntervalRelation(this, (TimeInterval)other) == EQUALS;
  }

  @Override
  public final int hashCode() {
    int result = 0;
    if (getBegin() != null) {
      result += getBegin().hashCode();
    }
    if (getEnd() != null) {
      result += getEnd().hashCode();
    }
    /* since we need to be consistent with equals, we can only take into account properties
       that are used by allenRelation, i.e., begin and end */
    return result;
  }

  @Override
  public String toString() {
    return "[" + toString(getBegin()) + ", " + toString(getEnd()) + "[\u0394(" + toStringDuration() + ")"; // \u0394 is Greek capital delta
  }

  private static String toString(Date d) {
    return d == null ? "|?|" : d.toString();
  }

  private String toStringDuration() {
    Duration d = getDuration();
    return d == null ? "-?-" : d.toRoughString();
  }

  public final Date determinateBegin(Date stubBegin) {
    return getBegin() != null ? getBegin() : stubBegin;
  }

  public final Date determinateEnd(Date stubEnd) {
    return getEnd() != null ? getEnd() : stubEnd;
  }

}

File: dotnet/Value/trunk/src/I/Time/Interval/AbstractTimeInterval.cs
/*<license>
Copyright 2011 - $Date: 2008-11-06 15:27:53 +0100 (Thu, 06 Nov 2008) $ by PeopleWare n.v..

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
</license>*/

#region Using

using System;
using System.Diagnostics.Contracts;

#endregion

namespace PPWCode.Value.I.Time.Interval
{
    /// <summary>
    /// General supporting code for time interval implementations.
    /// </summary>
    public abstract class AbstractTimeInterval : ITimeInterval /* MUDO , AbstractImmutableValue */
    {
        #region Abstract members of ITimeInterval

        public abstract DateTime? Begin { get; }
        public abstract DateTime? End { get; }
        public abstract TimeSpan? Duration { get; }
        public abstract ITimeInterval Determinate(DateTime? stubBegin, DateTime? stubEnd);

        #endregion

        #region Implemented members of ITimeInterval

        public override sealed bool Equals(object obj)
        {
            // F*CK
            Contract.Ensures(Contract.Result<bool>() ==
                ((obj == null || GetType() != obj.GetType())
                    ? false
                    : (TimeIntervalRelation.MostCertainTimeIntervalRelation(this, (ITimeInterval)obj) == TimeIntervalRelation.EQUALS)));

            return this.Equals(obj as ITimeInterval);
        }

        /// <summary>
        /// Equals as defined by <see cref="TimeIntervalRelation.TimeIntervalRelation"/>
        /// and <see cref="TimeIntervalRelation.EQUALS"/>.
        /// </summary>
        /// <remarks>
        /// See the
        //  <a href="http://go.microsoft.com/fwlink/?LinkID=85237">full list of guidelines</a>
        //  and also the guidance for <a href="http://go.microsoft.com/fwlink/?LinkId=85238">operator==</a>.
        /// </remarks>
        public bool Equals(ITimeInterval other)
        {
            Contract.Ensures(Contract.Result<bool>() == 
                ((other == null)
                    ? false
                    : (TimeIntervalRelation.MostCertainTimeIntervalRelation(this, other) == TimeIntervalRelation.EQUALS)));

            return (other != null) && GetType().IsInstanceOfType(other)
                   && (TimeIntervalRelation.MostCertainTimeIntervalRelation(this, other)
                       == TimeIntervalRelation.EQUALS);
        }

        /// <summary>
        /// Equality as defined by <see cref="TimeIntervalRelation.TimeIntervalRelation"/>
        /// and <see cref="TimeIntervalRelation.EQUALS"/>.
        /// </summary>
        /// <remarks>
        /// See the
        //  <a href="http://go.microsoft.com/fwlink/?LinkID=85237">full list of guidelines</a>
        //  and also the guidance for <a href="http://go.microsoft.com/fwlink/?LinkId=85238">operator==</a>.
        /// </remarks>
        public static bool operator ==(AbstractTimeInterval ati1, AbstractTimeInterval ati2)
        {
            // NO CONTRACT HERE, TO AVOID INFINITE RECURSION
            //Contract.Ensures(Contract.Result<bool>() == (ati1 == null ? ati2 == null : ati1.Equals(ati2)));

            return SaveIsNull(ati1) ? SaveIsNull(ati2) : ati1.Equals(ati2);
        }

        /// <summary>
        /// Equality as defined by <see cref="TimeIntervalRelation.TimeIntervalRelation"/>
        /// and <see cref="TimeIntervalRelation.EQUALS"/>.
        /// </summary>
        /// <remarks>
        /// See the
        //  <a href="http://go.microsoft.com/fwlink/?LinkID=85237">full list of guidelines</a>
        //  and also the guidance for <a href="http://go.microsoft.com/fwlink/?LinkId=85238">operator==</a>.
        /// </remarks>
        public static bool operator !=(AbstractTimeInterval ati1, AbstractTimeInterval ati2)
        {
            Contract.Ensures(Contract.Result<bool>() == (! (ati1 == ati2)));

            return (SaveIsNull(ati1) && SaveIsNotNull(ati2)) || (SaveIsNotNull(ati1) && (! ati1.Equals(ati2)));
        }

        /// <summary>
        /// Compare to null without using the overridden operator.
        /// </summary>
        private static bool SaveIsNull(AbstractTimeInterval ati)
        {
            object atiAsObject = ati;
            return atiAsObject == null;            
        }

        /// <summary>
        /// Compare to null without using the overridden operator.
        /// </summary>
        private static bool SaveIsNotNull(AbstractTimeInterval ati)
        {
            object atiAsObject = ati;
            return atiAsObject != null;
        }

        public override sealed int GetHashCode()
        {
            int result = 0;
            if (Begin != null)
            {
                result += Begin.GetHashCode();
            }
            if (End != null)
            {
                result += End.GetHashCode();
            }
            /* since we need to be consistent with equals, we can only take into account properties
               that are used by MostCertainTimeIntervalRelation, i.e., begin and end */
            return result;
        }

        public DateTime? DeterminateBegin(DateTime? stubBegin)
        {
            return Begin ?? stubBegin;
        }

        public DateTime? DeterminateEnd(DateTime? stubEnd)
        {
            return End ?? stubEnd;
        }

        public override string ToString()
        {
            return "[" + ToString(Begin) + ", " + ToString(End) + "[\u0394(" + DurationAsString + ")"; // \u0394 is Greek capital delta
        }

        private static string ToString(DateTime? d)
        {
            return d == null ? "|?|" : d.ToString();
        }

        private string DurationAsString
        {
            get
            {
                TimeSpan? ts = Duration;
                return ts == null ? "-?-" : ts.ToString();
            }
        }

        #endregion
    }
}